### PR TITLE
warm .m2 cache in prebuild

### PR DESCRIPTION
**Pull request by frenchy64**
> **Epic** #
> Close #
> Related #

<a name="qa">[§](#qa)</a> QA
============================

No QA is needed.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: one-line description for internal eyes only.
public: one line description for public release notes.
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -88,42 +101,47 @@ jobs:
...
+          path: |
+            ~/bin/lein
+            ~/.lein
+          key: ctia-lein-cache-${{ env.ACTIONS_CACHE_VERSION }}-${{ env.LEIN_VERSION }}
+      - name: Install Leiningen
+        if: steps.maven-cache.outputs.cache-hit != 'true' && steps.lein-cache.outputs.cache-hit != 'true'
+        run: |
+          cd "${BIN_PATH}"
+          # poor man's travis_retry
+          for i in {1..${ACTIONS_RETRY_TIMES}}; do curl -sLO "https://raw.githubusercontent.com/technomancy/leiningen/${LEIN_VERSION}/bin/lein" && break; done
```
Comment (vemv) [inline .github/workflows/pr.yml:137]: Merely curious, why are these Lein commands flaky?

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -192,22 +210,21 @@ jobs:
...
-            chmod a+x lein
-            lein version
-          fi
-
+          cd "${BIN_PATH}"
+          # poor man's travis_retry
+          for i in {1..${ACTIONS_RETRY_TIMES}}; do curl -sLO "https://raw.githubusercontent.com/technomancy/leiningen/${LEIN_VERSION}/bin/lein" && break; done
+          chmod a+x lein
+          # poor man's travis_retry
+          for i in {1..${ACTIONS_RETRY_TIMES}}; do lein version && break; done
```
Comment (vemv) [inline .github/workflows/pr.yml:227]: Probably you've heard of https://github.com/DeLaGuardo/setup-clojure ?

Might save some LOCs. Using it sha-versioned would constrain risks.

**Review by vemv — COMMENTED** [review]
Looks reasonable!

Is it possible to test out these changes in some sample branches, exercising the described edge cases?

File: .github/workflows/pr.yml (hunk 1 above)
Comment (frenchy64) [reply .github/workflows/pr.yml:137]: I've never observed lein downloads as flaky, these are looped mainly because it's in the category of "downloading stuff".

File: .github/workflows/pr.yml (hunk 2 above)
Comment (frenchy64) [reply .github/workflows/pr.yml:227]: Haven't had a great experience with that action in terms of caching. I can't see an indication that's improved, but IIRC caching just "didn't work".

**Comment by frenchy64** [comment]
I was thinking the same thing. It turned out to be somewhat tricky to remove the caches when basing a PR on this branch, so I directly "disabled" the caches in the commit `workaround cache eviction`, which exercised most of those edge cases except the strange ones like a fanned-out job failing to pick up a just-warmed cache.

**Review by vemv — APPROVED** [review]
LGTM!

**Review by ereteog — COMMENTED** [review]
@frenchy64 Before merging it, I would like to talk to you about this PR and the course of action to tackle the recent build problems that started about 3 weeks ago. I think that we need a global plan because it's like the test / build just reached a breaking point.
IMO we need first to identify first the real cause of the slow tests.
